MinMaxGame.java

```diff
@@ -1,99 +1,99 @@
+		ValueWithMove move = getBestMove();
+		lastMoveY=move.bestMoveY;
+		lastMoveX=move.bestMoveX;
+	
+		assert table[lastMoveY][lastMoveX] == 0;
+		table[lastMoveY][lastMoveX] = 2;
+		movesCompleted += 1;
+		if (winOnLastMove()) {
+			scoreBoard[1] += 1;
 			return 2;
+		}
 		else if (isFull())
 			return 3;
-		else
+		else {
+			activeAgent = 1;
 			return 0;
+		}
 	}
-	protected int[] getValue(int[][] table, int lastMoveY, int lastMoveX, 
-								int player, int depth) {
-		// Opponent makes move (lastMoveY, lastMoveX) leading to table. 
-		// RETURNS: 
-		//    First entry:  The value of the position for player.
-		//    Second entry: depth + The lenght of the resulting game assuming optimal play.
-		// 
-		// NOTE: opponent = 3 - player. The second coordinate is needed so that the 
-		// computer player wants to win as quickly as possible or survive as long 
-		// as it can.
+	protected ValueWithMove getBestMove() {
+		// Getting best move at a current state of the table.
+		// RETURNS: Value of position and best move realizing that value.
+		//    
+		// NOTE: opponent = 3 - player. 
+		//
+		// Return the value if the game is over after opponents move.
+		// In this case (-1,-1) for best move in place of null.
 
-		boolean lost = (Game.whoWon(table, lastMoveY, lastMoveX, 3-player, lengthToWin) 
-																			== 3-player);
-		if (lost) 
-			return new int[] {-1, depth};
-		else if (Game.isFull(table))
-			return new int[] {0, depth};
+		int tempAgent = activeAgent;
+		int tempY = lastMoveY;
+		int tempX = lastMoveX;
+		
+		activeAgent = 3 - tempAgent;
+		if (winOnLastMove()) {
+			return new ValueWithMove(-1, movesCompleted, -1, -1);
+		}
+		else if (isFull()) {
+			return new ValueWithMove(0, movesCompleted, -1, -1);
+		}
 		else {
-			int[] bestValue = new int[] {-10, depth};
-			int[] value = new int[2];
+			ValueWithMove move;
+			ValueWithMove bestMove = new ValueWithMove(-2,0, -1, -1);
 			for(int i = 0; i < table.length; i++) {
 				for(int j = 0; j < table[i].length; j++) {
 					if (table[i][j] == 0) {
 						// Trying out move (i,j)
-						table[i][j] = player;
-						value = getValue(table, i, j, 3-player, depth + 1);
+						lastMoveY = i;
+						lastMoveX = j;
+						table[i][j] = tempAgent;
+						activeAgent = 3 - tempAgent;
+						movesCompleted += 1;
+
+						// Evaluating the value of the action
+						move = getBestMove();
+						move.flipValue();
+
+						// Undoing tried move
 						table[i][j] = 0;
-						if ((-value[0] == bestValue[0] && bestValue[0] == 1 && 
-														value[1] < bestValue[1]) ||
-							(-value[0] == bestValue[0] && bestValue[0] == -1 && 
-														value[1] > bestValue[1]) ||
-							(-value[0] > bestValue[0])) {
-							
-							bestValue[0] = -value[0];
-							bestValue[1] = value[1];
-						}
+						movesCompleted -= 1;
+
+						// Cosmetics on the move
+						move.bestMoveY = i;
+						move.bestMoveX = j;
+
+						// Comparing to current best move
```

Comment:
Don't need the comment as the line of code is self explanatory

Reply:
Done